src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs

```diff
@@ -239,5 +339,223 @@ private CspParameters GetPrivateKeyCsp()
+            {
+                // Well-known CNG providers, keySpec is 0.
+                return 0;
+            }
+
+            const int NTE_BAD_KEYSET = unchecked((int)0x80090016);
+
+            try
+            {
+                CngKeyOpenOptions options = machineKey ? CngKeyOpenOptions.MachineKey : CngKeyOpenOptions.None;
+
+                using (CngKey.Open(keyName, provider, options))
+                {
+                    // It opened with keySpec 0, so use keySpec 0.
+                    return 0;
+                }
+            }
+            catch (CryptographicException e)
+            {
+                Debug.Assert(
+                    e.HResult == NTE_BAD_KEYSET,
+                    $"CngKey.Open had unexpected error: 0x{e.HResult:X8}: {e.Message}");
+
+                CspParameters cspParameters = new CspParameters
+                {
+                    ProviderName = provider.Provider,
+                    KeyContainerName = keyName,
+                    Flags = CspProviderFlags.UseExistingKey,
+                    KeyNumber = (int)KeyNumber.Signature,
+                };
+
+                if (machineKey)
+                {
+                    cspParameters.Flags |= CspProviderFlags.UseMachineKeyStore;
+                }
+
+                if (algorithmGroup == CngAlgorithmGroup.Rsa)
+                {
+                    // Try the AT_SIGNATURE spot in each of the 4 RSA provider type values,
+                    // ideally one of them will work.
+                    const int PROV_RSA_FULL = 1;
+                    const int PROV_RSA_SIG = 2;
+                    const int PROV_RSA_SCHANNEL = 12;
+                    const int PROV_RSA_AES = 24;
+
+                    // These are ordered in terms of perceived likeliness, given that the key
+                    // is AT_SIGNATURE.
+                    int[] provTypes =
+                    {
+                        PROV_RSA_FULL,
+                        PROV_RSA_AES,
+                        PROV_RSA_SCHANNEL,
+
+                        // Nothing should be PROV_RSA_SIG, but if everything else has failed,
+                        // just try this last thing.
+                        PROV_RSA_SIG,
+                    };
+
+                    foreach (int provType in provTypes)
+                    {
+                        cspParameters.ProviderType = provType;
+
+                        try
+                        {
+                            using (new RSACryptoServiceProvider(cspParameters))
+                            {
+                                return cspParameters.KeyNumber;
+                            }
+                        }
+                        catch (CryptographicException)
+                        {
+                        }
+                    }
+
+                    Debug.Fail("RSA key did not open with KeyNumber 0 or AT_SIGNATURE");
+                }
+                else if (algorithmGroup == CngAlgorithmGroup.Dsa)
```

Comment:
Yes, that's how they passed the algorithmGroup value.  I'm not sure what the implication is?  (This is part of a common codepath where we've already reduced to CngKey, and we only try unwrapping the CAPI-ness if CNG doesn't open it without the KeySpec value being explicit)